### PR TITLE
Reformat Validator Election

### PR DESCRIPTION
**Pull request by Kianenigma**
This gives us: 

- A much neater API for how the election is working 
  - Which does open the way for removing the backward vote-currency convert.
- Reduce the complexity of data pre-processing to make it non-exploitable. https://github.com/paritytech/substrate/compare/kiz-reformat-election?expand=1#diff-ba0a45203e7c14aedb758fe4d3653b35R137
  - And improves the accuracy actually.
- Add a mini post-processing to compensate for stakes being loss. This is WIP and any input is appreciated https://github.com/paritytech/substrate/compare/kiz-reformat-election?expand=1#diff-ba0a45203e7c14aedb758fe4d3653b35R250.
- Improve testing.

### CODE REVIEW

File: srml/staking/src/lib.rs [hunk 1]
```diff
@@ -925,20 +932,69 @@ impl<T: Trait> Module<T> {
...
+			// helper closure.
+			let to_balance = |b: ExtendedBalance| <T::CurrencyToVote as Convert<ExtendedBalance, BalanceOf<T>>>::convert(b);
+			let to_votes = |b: BalanceOf<T>| <T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b) as ExtendedBalance;
+
+			// The return value of this is safe to be converted to u64.
+			// The original balance, `b` is within the scope of u64. It is just extended to u128
+			// to be properly multiplied by a ratio, which will lead to another value
+			// less than u64 for sure. The result can then be safely passed to `to_balance`.
+			// For now the backward convert is used. A simple `TryFrom<u64>` is also safe.
+			let ratio_of = |b, p| (p as ExtendedBalance) * to_votes(b) / ACCURACY;
```
Comment (joepetrowski) [inline srml/staking/src/lib.rs]: Should we use `checked_mul()` to be safe?

File: srml/staking/src/phragmen.rs [hunk 2]
```diff
@@ -16,43 +16,35 @@
...
-}
+pub type ExtendedBalance = u128;
+
+// this is only used while creating the candidate score. Due to reasons explained below
+// The more accurate this is, the less likely we choose a wrong candidate.
+const SCALE_FACTOR: ExtendedBalance = u32::max_value() as ExtendedBalance + 1;
+// These are used to expose a fixed accuracy to the caller function. The bigger they are,
+// the more accurate we get, but the more likely it is for us to overflow. The case of overflow
+// is handled but accuracy will be lost. 32 or 16 are reasonable values.
+pub const ACCURACY: ExtendedBalance = u32::max_value() as ExtendedBalance + 1;
```
Comment (joepetrowski) [inline srml/staking/src/phragmen.rs]: Should a `pub` value have docs (`///`)?

File: srml/staking/src/phragmen.rs [hunk 3]
```diff
@@ -16,43 +16,35 @@
...
-/// Might be deprecated.
-pub struct ElectionConfig<Balance: HasCompact> {
-	/// Perform equalize?.
-	pub equalize: bool,
-	/// Number of equalize iterations.
-	pub iterations: usize,
-	/// Tolerance of max change per equalize iteration.
-	pub tolerance: Balance,
-}
+pub type ExtendedBalance = u128;
```
Comment (joepetrowski) [inline srml/staking/src/phragmen.rs]: If this is `pub`/used elsewhere, should there be some kind of warning, like "must be u64 or greater" (and any other restrictions) since it is used below with `u32::max_value() as ExtendedBalance + 1`?

File: srml/staking/src/phragmen.rs [hunk 4]
```diff
@@ -77,14 +69,10 @@ pub struct Edge<AccountId> {
 	who: AccountId,
 	/// Load of this vote.
 	load: Fraction,
-	/// Final backing stake of this vote.
-	backing_stake: ExtendedBalance,
+	/// equal to `edge.load / nom.load`. Stored only to be used with post-processing.
```
Comment (joepetrowski) [inline srml/staking/src/phragmen.rs]:
```suggestion
	/// Equal to `edge.load / nom.load`. Stored only to be used with post-processing.
```

File: srml/staking/src/phragmen.rs [hunk 5]
```diff
@@ -77,14 +69,10 @@ pub struct Edge<AccountId> {
...
+	ratio: ExtendedBalance,
 	/// Index of the candidate stored in the 'candidates' vector
```
Comment (joepetrowski) [inline srml/staking/src/phragmen.rs]:
```suggestion
	/// Index of the candidate stored in the 'candidates' vector.
```

File: srml/staking/src/phragmen.rs [hunk 6]
```diff
@@ -93,70 +81,84 @@ pub struct Edge<AccountId> {
 ///
 /// Returns an Option of elected candidates, if election is performed.
 /// Returns None if not enough candidates exist.
+///
+/// The returned Option is a tuple consisting of:
+///   - The list elected candidates
```
Comment (joepetrowski) [inline srml/staking/src/phragmen.rs]:
```suggestion
///   - The list of elected candidates.
```

File: srml/staking/src/phragmen.rs [hunk 7]
```diff
@@ -207,162 +212,180 @@ pub fn elect<T: Trait + 'static, FV, FN, FS>(
...
+		let max_stake = backing_backed_stake
 			.iter()
 			.max()
 			.expect("vector with positive length will have a max; qed");
-		let min_stake = *backed_stakes
-			.iter()
+		let min_stake = backed_stakes_iter
 			.min()
-			.expect("vector with positive length will have a min; qed");
+			.expect("iterator with positive length will have a min; qed");
```
Comment (joepetrowski) [inline srml/staking/src/phragmen.rs]: Yes, but you did not check that `backed_stakes_iter.len() > 0` before entering this block. Seems that to be certain you'd have to check that the input `elected_edges_balance` is longer than zero before passing it to this function, so probably best to put the check within the function, no?

File: srml/staking/src/phragmen.rs [hunk 8]
```diff
@@ -207,162 +212,180 @@ pub fn elect<T: Trait + 'static, FV, FN, FS>(
...
 	}
-	Some(elected_candidates)
+	Some((elected_candidates, assigned))
 }
 
 /// Performs equalize post-processing to the output of the election algorithm
 /// This function mutates the input parameters, most noticeably it updates the exposure of
 /// the elected candidates.
+///
 /// The return value is to tolerance at which the function has stopped.
```
Comment (joepetrowski) [inline srml/staking/src/phragmen.rs]: There is no more return value. Does this break API?

File: srml/staking/src/lib.rs [hunk 9]
```diff
@@ -394,12 +394,19 @@ type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as system::Trait>::Ac
...
+type Assignment<T> = (<T as system::Trait>::AccountId, ExtendedBalance, BalanceOf<T>);
+type ExpoMap<T> = BTreeMap::<<T as system::Trait>::AccountId, Exposure<<T as system::Trait>::AccountId, BalanceOf<T>>>;
+
 pub trait Trait: system::Trait + session::Trait {
 	/// The staking balance.
 	type Currency: LockableCurrency<Self::AccountId, Moment=Self::BlockNumber>;
 
 	/// Convert a balance into a number used for election calculation.
 	/// This must fit into a `u64` but is allowed to be sensibly lossy.
+	/// TODO: #https://github.com/paritytech/substrate/pull/2205#issuecomment-483551470
```
Comment (bkchr) [inline srml/staking/src/lib.rs]: Please link to an issue.

File: srml/staking/src/phragmen.rs (hunk 2 above)
Comment (bkchr) [reply srml/staking/src/phragmen.rs]: Yes

File: srml/staking/src/lib.rs (hunk 1 above)
Comment (Kianenigma) [reply srml/staking/src/lib.rs]: The comment above exactly explains why this is actually safe. Will change to saturating just in case.

If the comment is not sufficient to explain why, I'd be happy to be further challenged.

`p` is a value between `0` and `ACCURACY`, while `b` is a normal `balance` type elevated to u128. Hence, `p / ACCURACY    * b` is indeed less than or equal to `b` and can be safely multiplied.

File: srml/staking/src/phragmen.rs (hunk 3 above)
Comment (Kianenigma) [reply srml/staking/src/phragmen.rs]: Good point. Yet, this is not something configurable. In fact it is set-in-stone since the module is asking for an explicit `Convert<Balance, u64>`, so this will always be equal to u128. 

And, this is really just a bit of syntax sugar otherwise the intention is to _always_ use u128.

File: srml/staking/src/phragmen.rs (hunk 7 above)
Comment (Kianenigma) [reply srml/staking/src/phragmen.rs]: So `backed_stakes_iter` is a superset of `backing_backed_stake` (see the filter applied to both) which is checked to have a positive length. I believe this is safe enough to not add another condition.

File: srml/staking/src/phragmen.rs (hunk 8 above)
Comment (Kianenigma) [reply srml/staking/src/phragmen.rs]: Good catch to update the doc but nothing breaking; these are just internal functions.

File: srml/staking/src/lib.rs (hunk 1 above)
Comment (joepetrowski) [reply srml/staking/src/lib.rs]: But your line is `p * b / A`. I don't know how the CPU evaluates that; is it possible for `p * b` to overflow before division even occurs? Of course, if `p` and `b` are both u64 cast to u128, it's a total edge case that it comes close to overflow (u64::max * u64::max), but if there is no/little overhead for saturating then I don't see the harm.

File: srml/staking/src/tests.rs [hunk 10]
```diff
@@ -2026,38 +1962,66 @@ fn large_scale_test() {
...
+	})
+}
 
-		// For manual inspection
-		println!("Validators are {:?}", Session::validators());
-		println!("Validators are {:#?}",
-			Session::validators().iter().map(|v| (v.clone(), Staking::stakers(v-1)) ).collect::<Vec<(u64, Exposure<u64, u64>)>>());
+#[test]
+fn phragmen_large_scale_test_2() {
+	with_externalities(&mut ExtBuilder::default()
```
Comment (gavofyork) [inline srml/staking/src/tests.rs]: incorrect indentation (should be an extra level)

File: srml/staking/src/tests.rs [hunk 11]
```diff
@@ -2026,38 +1962,66 @@ fn large_scale_test() {
...
+	.nominate(false)
```
Comment (gavofyork) [inline srml/staking/src/tests.rs]: incorrect indentation (should be an extra level)

**Comment by gavofyork** [comment]
Needs a deep review of the logic by somebody before merge as it appears to be essentially a rewrite.

**Comment by AlistairStewart** [comment]
The logic looks fine on a first read-through.

I think the complexity of candidate look ups is down from number of nominators * number of candiates to number of candidates^2, which is still quadratic. Instead of ever looking up the candidates, you could populate the B-tree by iterating through them, which would get the complexity down to nearly linear.

For equalising, it's worth noting that for the  pyython, where I equalised random nominators and then iterating once, one iteration sufficed for the random examples. So putting the equalising step on by default with one or two iterations would give a decent gain, unless it is noticebly slow.

**Comment by Kianenigma** [comment]
> Needs a deep review of the logic by somebody before merge as it appears to be essentially a rewrite.

Agreed. Already asked Al to have a look. 

Regarding it being a rewrite: I should mention that the tests that examine detailed values (mostly a pain to maintain) of the election, stake, and reward, etc. stayed unchanged in this PR (except for minor tweak to values like +=1, 10 which is due to accuracy change). I take that as a good sign that it is not a major breaking change. 

A recap of the important stuff that I'd rather not be overlooked: 
- Where we are forced to saturate and there is no other way since being exposed by more balance than the maximum amount of balance does not make any sense: https://github.com/paritytech/substrate/pull/2406/files/2fbd2ac901861dee5d0f4acb974d149d2ea00df5#diff-b5a024b4f4ae96b0bfc0681acae56709R980
- My little made up post-processing to compensate for weight loss due to accuracy (e.g. `[33%, 33%, 33%] -> 1 is burnt`). Logs show that at least during most tests the amount of `diff` to be compensated is less than the number of votes per nominator. https://github.com/paritytech/substrate/pull/2406/files/2fbd2ac901861dee5d0f4acb974d149d2ea00df5#diff-ba0a45203e7c14aedb758fe4d3653b35R247

**Comment by Kianenigma** [comment]
> I think the complexity of candidate look ups is down from number of nominators * number of candiates to number of candidates^2, which is still quadratic. Instead of ever looking up the candidates, you could populate the B-tree by iterating through them, which would get the complexity down to nearly linear.

Ahh I see. Such an obvious error of mine... Fixed.

> For equalising, it's worth noting that for the pyython, where I equalised random nominators and then iterating once, one iteration sufficed for the random examples. So putting the equalising step on by default with one or two iterations would give a decent gain, unless it is noticebly slow.

I would leave that for 1) potentially off-chain 2) after an initial complexity analysis of the already existing code before exposing it more.


I'd still like to hear some feedback on the points linked in my previous comment.

**Comment by gavofyork** [comment]
@AlistairStewart *bump* anything else to add on Kian's changes?

**Comment by Kianenigma** [comment]
In the meantime and in light of preventing this PR to become too stale I've had one more look at the code and minor fixes + making issues for some of the gray zones.

**Comment by gavofyork** [comment]
@Kianenigma could you chase @AlistairStewart to sign off on the other things you were unsure about here?

**Comment by Kianenigma** [comment]
> @Kianenigma could you chase @AlistairStewart to sign off on the other things you were unsure about here?

had pinged him a few times already. Will asked again.

**Comment by AlistairStewart** [comment]
This still looks good. The complexity is fixed. The two things you wanted to highlight both look fine. For the first, I'm sure that having balances that add up to over the maximum balance might break other things in unexpected ways, but I'd agree that the only thing to do here is to saturate which should be ok.
